### PR TITLE
Port h3 features/encoder to Pytorch

### PR DESCRIPTION
**Pull request by justinxzhao**

### CODE REVIEW

File: ludwig/encoders/h3_encoders.py [hunk 1]
```diff
@@ -180,7 +163,9 @@ def __init__(
         )
 
         logger.debug('  FCStack')
+        self.fc_size = fc_size
```
Comment (ShreyaR) [inline ludwig/encoders/h3_encoders.py:166]: Is this used anywhere?

File: ludwig/encoders/h3_encoders.py [hunk 2]
```diff
@@ -368,105 +307,95 @@ def __init__(
...
-            activity_regularizer=None,
-            dropout=0.0,
-            recurrent_dropout=0.0,
-            reduce_output='last',
-            **kwargs
+            embedding_size: int = 10,
+            embeddings_on_cpu: bool = False,
+            num_layers: int = 1,
+            state_size: int = 10,
+            cell_type: int = 'rnn',
```
Comment (ShreyaR) [inline ludwig/encoders/h3_encoders.py:359]: Seems like this should be a `str` type

File: ludwig/modules/recurrent_modules.py [hunk 3]
```diff
@@ -35,33 +35,16 @@
...
-            weights_regularizer=None,
-            recurrent_regularizer=None,
-            bias_regularizer=None,
-            activity_regularizer=None,
-            # kernel_constraint=kernel_constraint,
-            # recurrent_constraint=recurrent_constraint,
-            # bias_constraint=bias_constraint,
-            dropout=0.0,
-            recurrent_dropout=0.0,
+            input_size: int = None,
```
Comment (ShreyaR) [inline ludwig/modules/recurrent_modules.py:38]: Could you add a unit test for `RecurrentStack` in `tests/ludwig/modules/test_recurrent_modules.py`?

File: ludwig/modules/embedding_modules.py [hunk 4]
```diff
@@ -409,9 +414,8 @@ def __init__(
         )
 
         if embedding_regularizer:
-            embedding_regularizer_obj = tf.keras.regularizers.get(
-                embedding_regularizer)
-            self.add_loss(lambda: embedding_regularizer_obj(self.embeddings))
+            self.add_loss(lambda: reg_loss(
+                self.embeddings, embedding_regularizer))
```
Comment (ShreyaR) [inline ludwig/modules/embedding_modules.py:418]: Could you also add a test for `EmbedSequence` under `tests/ludwig/modules/test_embedding_modules.py`?

**Review by ShreyaR — COMMENTED** [review]
Looks great! Added some comments for tests, and some minor comments

File: ludwig/encoders/h3_encoders.py (hunk 2 above)
Comment (justinxzhao) [reply ludwig/encoders/h3_encoders.py:359]: Good catch! Done.

File: ludwig/modules/embedding_modules.py (hunk 4 above)
Comment (justinxzhao) [reply ludwig/modules/embedding_modules.py:418]: Done.

File: ludwig/modules/recurrent_modules.py (hunk 3 above)
Comment (justinxzhao) [reply ludwig/modules/recurrent_modules.py:38]: Done.

File: tests/ludwig/modules/test_recurrent_modules.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import logging
+import torch
+
+from ludwig.modules import recurrent_modules
+
+logger = logging.getLogger(__name__)
+
+
+def test_recurrent_stack():
```
Comment (ShreyaR) [inline tests/ludwig/modules/test_recurrent_modules.py:9]: You could potentially consolidate `test_recurrent_stack` and `test_recurrent_stack_NoSequenceSize` into a single test by using `pytest.mark.parameterize`.  E.g.
```
@pytest.mark.parameterize('sequence_size', [19, None])
```

This is probably good for now, but might be useful for future tests.

**Review by ShreyaR — APPROVED** [review]
Looks great! Feel free to merge once @w4nderlust approves

File: ludwig/modules/embedding_modules.py [hunk 6]
```diff
@@ -79,9 +79,8 @@ def embedding_matrix(
 
     elif representation == 'sparse':
         embedding_size = vocab_size
-        embeddings = torch.tensor(
-            get_initializer('identity')([vocab_size, embedding_size]),
-            requires_grad=False)
+        embeddings = get_initializer('identity')(
+            [vocab_size, embedding_size]).clone().detach().requires_grad_(False)
```
Comment (w4nderlust) [inline ludwig/modules/embedding_modules.py:83]: curious here, why are clone and detach needed and why this different way of setting requires grad to false?

File: ludwig/encoders/h3_encoders.py [hunk 7]
```diff
@@ -42,70 +48,49 @@ class H3Embed(H3Encoder):
...
-                   is greater than 0).
-            :type regularize: Boolean
         """
         super().__init__()
         logger.debug(' {}'.format(self.name))
 
         self.embedding_size = embedding_size
         self.reduce_output = reduce_output
-        self.reduce_sequence = SequenceReducer(reduce_mode=reduce_output)
+        self.sum_sequence_reducer = SequenceReducer(reduce_mode=reduce_output)
```
Comment (w4nderlust) [inline ludwig/encoders/h3_encoders.py:93]: I prefer the old name, as which reduce operation to use is user specified, it could be something different than sum.

File: ludwig/modules/embedding_modules.py (hunk 6 above)
Comment (justinxzhao) [reply ludwig/modules/embedding_modules.py:83]: This was a warning/recommendation from pytorch:

```
UserWarning: To copy construct from a tensor, it is recommended to use source
Tensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True),
 rather than torch.tensor(sourceTensor)
```

It looks like it's slightly more efficient, though also less readable:
https://stackoverflow.com/questions/55266154/pytorch-preferred-way-to-copy-a-tensor

File: ludwig/encoders/h3_encoders.py (hunk 7 above)
Comment (justinxzhao) [reply ludwig/encoders/h3_encoders.py:93]: Done.

Comment (justinxzhao) [reply ludwig/encoders/h3_encoders.py]: I kept the other sequence reducer in this file as sum_sequence_reducer since that one seems to be configured strictly with reduce_mode=sum.

File: ludwig/modules/embedding_modules.py (hunk 6 above)
Comment (justinxzhao) [reply ludwig/modules/embedding_modules.py:83]: I like the second option. Yeah it's not clear why in the original code we had to copy the tensor returned by get_initializer(...) into another tensor.
